packages/shorebird_cli/lib/src/executables/devicectl/devicectl.dart

```diff
@@ -0,0 +1,311 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:collection/collection.dart';
+import 'package:io/io.dart';
+import 'package:json_path/json_path.dart';
+import 'package:path/path.dart' as p;
+import 'package:scoped/scoped.dart';
+import 'package:shorebird_cli/src/executables/devicectl/apple_device.dart';
+import 'package:shorebird_cli/src/executables/devicectl/nserror.dart';
+import 'package:shorebird_cli/src/logger.dart';
+import 'package:shorebird_cli/src/process.dart';
+import 'package:shorebird_cli/src/third_party/flutter_tools/lib/flutter_tools.dart';
+import 'package:shorebird_code_push_client/shorebird_code_push_client.dart';
+
+typedef BundleId = String;
+
+/// {@template devicectl_exception}
+/// Thrown when a [Devicectl] command fails.
+/// {@endtemplate}
+class DevicectlException implements Exception {
+  /// {@macro devicectl_exception}
+  DevicectlException({
+    required this.message,
+    this.underlyingException,
+  });
+
+  /// A message describing this exception.
+  final String message;
+
+  /// The exception that caused this exception to be thrown, if any.
+  final Exception? underlyingException;
```

Comment:
Or should this just be an `Object?` to align with `catch`?

Reply:
Good call, this also lets us avoid the `on Exception`s below. Done